Content type is not set when HTB is used with a small buffer

References https://github.com/klauspost/compress/pull/762#discussion_r1388355061

It seems that the `Content-Type` header is not set when the size of the response body is smaller than the jitter entropy buffer, which doesn't seem like normal behavior at a first glance. As requested, please find a reproduction below:

```go
func TestContentTypeDetectWithJitter(t *testing.T) {
	t.Parallel()
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		content := `<!DOCTYPE html>` + strings.Repeat("foo", 400)
		w.Write([]byte(content))
	})

	for _, tc := range []struct {
		name    string
		wrapper func(http.Handler) (http.Handler, error)
	}{
		{
			name: "no wrapping",
			wrapper: func(h http.Handler) (http.Handler, error) {
				return h, nil
			},
		},
		{
			name: "default",
			wrapper: func(h http.Handler) (http.Handler, error) {
				wrapper, err := NewWrapper()
				if err != nil {
					return nil, err
				}
				return wrapper(h), nil
			},
		},
		{
			name: "jitter, default buffer",
			wrapper: func(h http.Handler) (http.Handler, error) {
				wrapper, err := NewWrapper(RandomJitter(32, 0, false))
				if err != nil {
					return nil, err
				}
				return wrapper(h), nil
			},
		},
		{
			name: "jitter, small buffer",
			wrapper: func(h http.Handler) (http.Handler, error) {
				wrapper, err := NewWrapper(RandomJitter(32, DefaultMinSize, false))
				if err != nil {
					return nil, err
				}
				return wrapper(h), nil
			},
		},
	} {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			handler, err := tc.wrapper(handler)
			assertNil(t, err)

			req, resp := httptest.NewRequest(http.MethodGet, "/", nil), httptest.NewRecorder()
			req.Header.Add("Accept-Encoding", "gzip")

			handler.ServeHTTP(resp, req)

			assertEqual(t, "text/html; charset=utf-8", resp.Header().Get("Content-Type"))
		})
	}
}
```

On `go version go1.21.3 darwin/arm64` this outputs:
```
--- FAIL: TestContentTypeDetectWithJitter (0.00s)
    --- FAIL: TestContentTypeDetectWithJitter/jitter,_default_buffer (0.00s)
        /Users/adriansmares/projects/compress/gzhttp/compress_test.go:1716: want "text/html; charset=utf-8", got "application/x-gzip"
FAIL
FAIL	github.com/klauspost/compress/gzhttp	0.181s
FAIL
```

As mentioned in the original issue, I think the culprit is the jitter buffer size guard added here:
https://github.com/klauspost/compress/blob/847c1b439964030edb8aa504f4eeb67819f24bb4/gzhttp/compress.go#L127-L130

I cannot tell what was the original intent, but I would've expected that the size of the entropy pool won't affect the `Content-Type` logic.

cc @klauspost (and apologies for the original necropost).

--- a/gzhttp/compress.go
+++ b/gzhttp/compress.go
@@ -335,7 +335,16 @@ func (w *GzipResponseWriter) Close() error {
 			ce = w.Header().Get(contentEncoding)
 			cr = w.Header().Get(contentRange)
 		)
-		// fmt.Println(len(w.buf) == 0, len(w.buf) < w.minSize, len(w.Header()[HeaderNoCompression]) != 0, ce != "", cr != "", !w.contentTypeFilter(ct))
+		if ct == "" {
+			ct = http.DetectContentType(w.buf)
+
+			// Handles the intended case of setting a nil Content-Type (as for http/server or http/fs)
+			// Set the header only if the key does not exist
+			if _, ok := w.Header()[contentType]; w.setContentType && !ok {
+				w.Header().Set(contentType, ct)
+			}
+		}
+
 		if len(w.buf) == 0 || len(w.buf) < w.minSize || len(w.Header()[HeaderNoCompression]) != 0 || ce != "" || cr != "" || !w.contentTypeFilter(ct) {
 			// GZIP not triggered, write out regular response.
 			return w.startPlain()

--- a/gzhttp/compress_test.go
+++ b/gzhttp/compress_test.go
@@ -14,6 +14,7 @@ import (
 	"net/url"
 	"os"
 	"strconv"
+	"strings"
 	"testing"
 
 	"github.com/klauspost/compress/gzip"
@@ -1882,4 +1883,69 @@ func Test1xxResponses(t *testing.T) {
 
 	body, _ := io.ReadAll(res.Body)
 	assertEqual(t, gzipStrLevel(testBody, gzip.DefaultCompression), body)
+}
+
+func TestContentTypeDetectWithJitter(t *testing.T) {
+	t.Parallel()
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		content := `<!DOCTYPE html>` + strings.Repeat("foo", 400)
+		w.Write([]byte(content))
+	})
+
+	for _, tc := range []struct {
+		name    string
+		wrapper func(http.Handler) (http.Handler, error)
+	}{
+		{
+			name: "no wrapping",
+			wrapper: func(h http.Handler) (http.Handler, error) {
+				return h, nil
+			},
+		},
+		{
+			name: "default",
+			wrapper: func(h http.Handler) (http.Handler, error) {
+				wrapper, err := NewWrapper()
+				if err != nil {
+					return nil, err
+				}
+				return wrapper(h), nil
+			},
+		},
+		{
+			name: "jitter, default buffer",
+			wrapper: func(h http.Handler) (http.Handler, error) {
+				wrapper, err := NewWrapper(RandomJitter(32, 0, false))
+				if err != nil {
+					return nil, err
+				}
+				return wrapper(h), nil
+			},
+		},
+		{
+			name: "jitter, small buffer",
+			wrapper: func(h http.Handler) (http.Handler, error) {
+				wrapper, err := NewWrapper(RandomJitter(32, DefaultMinSize, false))
+				if err != nil {
+					return nil, err
+				}
+				return wrapper(h), nil
+			},
+		},
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			handler, err := tc.wrapper(handler)
+			assertNil(t, err)
+
+			req, resp := httptest.NewRequest(http.MethodGet, "/", nil), httptest.NewRecorder()
+			req.Header.Add("Accept-Encoding", "gzip")
+
+			handler.ServeHTTP(resp, req)
+
+			assertEqual(t, "text/html; charset=utf-8", resp.Header().Get("Content-Type"))
+		})
+	}
 }

